Setting trust_remote_code to True for HuggingFace datasets compatibility

veekaybee opened the pull request:
See this issue for context: https://github.com/EleutherAI/lm-evaluation-harness/issues/1135#issuecomment-1961928695

We'd like to be able to use the latest datasets version, 2.16 in lm-evaluation-harness. We currently [use 2.15](https://github.com/EleutherAI/lm-evaluation-harness/pull/1312) . 

In order to accommodate this, we'd like to:

1. add a `trust_remote_code` dataset kwarg to each dataset that requires it and pass through. [See notebook](https://gist.github.com/veekaybee/9fa165f5f41fc57d6e2ccd8c7644ca5d) for derivation of which datasets require it. 
```
['EleutherAI/mutual',
 'skt/kobest_v1',
 'EleutherAI/logiqa',
 'bigbio/pubmed_qa',
 'EleutherAI/race',
 'BigScienceBiasEval/crows_pairs_multilingual',
 'baber/logiqa2',
 'EleutherAI/arithmetic',
 'EleutherAI/asdiv',
 'EleutherAI/lambada_openai',
 'EleutherAI/wikitext_document_level',
 'EleutherAI/drop',
 'EleutherAI/hendrycks_math',
 'juletxara/xstory_cloze',
 'allenai/qasper',
 'EleutherAI/pile',
 'EleutherAI/sycophancy',
 'EleutherAI/hendrycks_ethics',
 'EleutherAI/unscramble',
 'EleutherAI/headqa',
 'skg/toxigen-data',
 'EleutherAI/coqa',
 'corypaik/prost']
```

3. Include `trust_remote_code` as True by default in the model constructor, to be overriden by -model_args when evaluating from the command line.

4. Respect the user's `HF_DATASETS_TRUST_REMOTE_CODE` if it exists. 

Related: https://github.com/EleutherAI/lm-evaluation-harness/pull/1467

cc @lhoestq fyi

veekaybee commented on lm_eval/api/task.py:
```diff
@@ -220,6 +221,8 @@ def __init__(
         self._config = TaskConfig({**config}) if config else TaskConfig()
 
         self._filters = [build_filter_ensemble("none", [["take_first", None]])]
+        self.trust_remote_code = os.environ.get("HF_DATASETS_TRUST_REMOTE_CODE", None)
```
does it make sense to pull from env variables based on how datasets sets it?

haileyschoelkopf reviewed (commented):
Thank you @veekaybee !

Upon thinking some more, I think actually, let's do the following (slightly differently to how I originally said it):

- keep `trust_remote_code = False` by default in model constructor
- have `trust_remote_code: True` in the YAMLs as you added *but only for datasets under the EleutherAI/ HF org or datasets in the default namespace (no org)* , the others should not get it set
- respect HF_DATASETS_TRUST_REMOTE_CODE env variable when it is set
- add a `--trust_remote_code` CLI flag, which, when set will: 1) cause `trust_remote_code=True` to be added to model_args string / passed to model constructor, 2) set the HF_DATASETS_TRUST_REMOTE_CODE env variable to true in __main__.py (or maybe in simple_evaluate()? we want running on remote-code requiring datasets to be not too annoying for people not using __main__.py)

How's this sound? basically, the PR as you've implemented it, but with an extra `--trust_remote_code` flag in the CLI, and with datasets not in the EAI or HF namespaces requiring this flag. Reasoning being that we'd like to support HF's initiative of making the OSS ecosystem more secure.

haileyschoelkopf commented:
> E   NotADirectoryError: [Errno 20] Not a directory: '/home/runner/.cache/huggingface/datasets/downloads/9d10351eefe83ab9887de1b307f40404b99de9ba10fed427d64faa36ae611778/HEAD_EN/train_HEAD_EN.json'

> The tests are failing at the download step and I'm guessing it's because of the environment variable

This looks like a separate error relating to HeadQA!

veekaybee commented:
@haileyschoelkopf Thank you! All of that makes sense and I think I've addressed all the comments, please take a look at the new implementation of the check for the trust_remote_code environment variable in main, and I'll check to see how tests are doing, as well.

veekaybee commented:
Getting new testing failures triggering from update of the YAML files I'm assuming, this time from `pile_arxiv.yaml`:

```
BuilderConfig 'pile_arxiv' not found. Available: ['all', 'enron_emails', 'europarl', 'free_law', 'hacker_news', 'nih_exporter', 'pubmed', 'pubmed_central', 'ubuntu_irc', 'uspto', 'github']
```

should the `task` be set to one of these? https://huggingface.co/datasets/EleutherAI/pile/blob/main/pile.py#L61

lhoestq commented:
AFAIK the `EleutherAI/pile` dataset has been taken down on its original host (the-eye) and is not accessible anymore

veekaybee commented:
In that case, I can take out the kwargs from that YAML file which I think is triggering the test errors -does it also make sense to delete/modify the task?

veekaybee commented:
@haileyschoelkopf this is ready for re-review! I removed the headqa task because it was erroring out on expected filetypes and on taking a look, the file pointers are to `pickle` files which I'd assume we want to avoid (there is a new set of parquet files that this points to but I'm not sure if we want to use those) https://huggingface.co/datasets/EleutherAI/headqa/discussions/1

veekaybee commented:
I'm also assuming that once we merge this we'll want to pin this to 2.17 https://github.com/EleutherAI/lm-evaluation-harness/pull/1312/files

veekaybee commented:
Hi all, I wanted to check on the status of this issue and if it was ok as-is, still needs more review, or is not needed anymore. Thanks!

haileyschoelkopf reviewed (commented):
I apologize for the delayed review @veekaybee ! Thank you very much for your work on this!

It looks good to me, and I think we can pin `datasets>=2.16` now! 

I've left some nits but otherwise LGTM!

veekaybee commented on lm_eval/api/task.py:
```diff
@@ -220,6 +221,7 @@ def __init__(
         self._config = TaskConfig({**config}) if config else TaskConfig()
 
         self._filters = [build_filter_ensemble("none", [["take_first", None]])]
+        self.cache_dir = os.environ.get("HF_DATASETS_CACHE", None)
```
Not required, just based on some discussion previously in the thread - I'll remove.

veekaybee commented:
Should be all addressed @haileyschoelkopf , thank you for taking a look!

haileyschoelkopf reviewed (approved):
Thank you for your work on this!
